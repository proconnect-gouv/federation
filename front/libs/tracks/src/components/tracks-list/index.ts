/* istanbul ignore file */

// declarative file
export * from './track-list.component';
