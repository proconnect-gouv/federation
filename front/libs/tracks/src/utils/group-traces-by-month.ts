import { DateTime } from 'luxon';
import { EnhancedTrack, TrackList, TracksConfig } from '../interfaces';

export function createUniqueGroupKeyFromTraceDate(
  trace: EnhancedTrack,
): number {
  // crée une clé unique pour un groupe
  // a partir de l'année et du mois
  const next = DateTime.fromObject({
    month: trace.datetime.month,
    year: trace.datetime.year,
  }).toMillis();
  return next;
}

export function groupTracksByMonth(
  options: TracksConfig,
  acc: TrackList[],
  trace: EnhancedTrack,
  index: number,
): TrackList[] {
  const isFirstTrace = index === 0;
  const previousGroup = isFirstTrace ? null : acc[acc.length - 1];
  const previousGroupKey = (previousGroup && previousGroup[0]) || null;

  const currentGroupKey = createUniqueGroupKeyFromTraceDate(trace);
  const isNotSameMonthYearGroup = currentGroupKey !== previousGroupKey;
  const shouldCreateNewTrackList = isFirstTrace || isNotSameMonthYearGroup;

  const nextTrackList = (
    !shouldCreateNewTrackList
      ? acc.pop()
      : [currentGroupKey, { label: null, traces: [] }]
  ) as TrackList;

  nextTrackList[1].label = !shouldCreateNewTrackList
    ? nextTrackList[1].label
    : trace.datetime.toFormat(options.LUXON_FORMAT_MONTH_YEAR);

  nextTrackList[1].traces = !shouldCreateNewTrackList
    ? [...((previousGroup && previousGroup[1].traces) || []), trace]
    : [trace];

  const next = [...acc, nextTrackList];
  return next;
}
