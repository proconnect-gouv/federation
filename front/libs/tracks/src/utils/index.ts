/* istanbul ignore file */

// declarative file
export * from './tracks.util';
