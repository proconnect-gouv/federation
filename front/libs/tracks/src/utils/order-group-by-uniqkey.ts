import { TrackList } from '../interfaces';

export function orderGroupByKeyAsc(groupA: TrackList, groupB: TrackList) {
  const uniqKeyA = groupA[0];
  const uniqKeyB = groupB[0];
  return uniqKeyB - uniqKeyA;
}
