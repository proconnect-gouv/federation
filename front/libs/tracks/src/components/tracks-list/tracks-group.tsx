import { ucfirst } from '@fc/common';
import { EnhancedTrack, TracksConfig } from '../../interfaces';
import { orderTracksByDateAsc } from '../../utils';
import { TrackCardComponent } from '../track-card/track-card.component';

export type TracksMonthGroupProps = {
  label: string;
  tracks: EnhancedTrack[];
  options: TracksConfig;
};

export const TracksGroupComponent = ({
  label,
  tracks,
  options,
}: TracksMonthGroupProps) => (
  <section className="mb40">
    <h6 className="pb12 mb16">
      <b>{ucfirst(label)}</b>
    </h6>
    {tracks.sort(orderTracksByDateAsc).map((track: EnhancedTrack) => (
      <TrackCardComponent key={track.trackId} track={track} options={options} />
    ))}
  </section>
);

TracksGroupComponent.displayName = 'TracksMonthGroupComponent';
