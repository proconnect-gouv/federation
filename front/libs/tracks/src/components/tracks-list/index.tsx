import axios from 'axios';
import { useCallback, useEffect, useRef, useState } from 'react';

import { TracksGroupComponent } from './tracks-group';
import {
  groupTracksByMonth,
  orderGroupByKeyAsc,
  transformTraceToEnhanced,
} from '../../utils';
import { TrackList, TracksConfig } from '../../interfaces';

export type TracksListComponentProps = {
  options: TracksConfig;
};

export const TracksListComponent = ({ options }: TracksListComponentProps) => {
  const isMounted = useRef(false);

  const [traces, setTraces] = useState([]);

  const responseErrorHandler = useCallback(() => {
    // @TODO
    // afficher une erreur lorsque le chargement
    // des traces ne s'est pas effectuÃ© -> @maxime
  }, []);

  const responseSuccessHandler = useCallback(({ data }) => {
    setTraces(data);
  }, []);

  useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;
      axios
        .get(options.API_ROUTE_TRACKS)
        .then(responseSuccessHandler)
        .catch(responseErrorHandler);
    }
  });

  return (
    <div className="mt40">
      {traces
        .map(transformTraceToEnhanced)
        .reduce(groupTracksByMonth.bind(null, options), [])
        .sort(orderGroupByKeyAsc)
        .map(([key, { label, traces: items }]: TrackList) => (
          <TracksGroupComponent
            key={key}
            label={label}
            traces={items}
            options={options}
          />
        ))}
    </div>
  );
};
