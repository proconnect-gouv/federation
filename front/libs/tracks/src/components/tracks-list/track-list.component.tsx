import { useApiGet } from '@fc/common';

import { TracksGroupComponent } from './tracks-group';
import {
  groupTracksByMonth,
  orderGroupByKeyAsc,
  transformTrackToEnhanced,
} from '../../utils';
import { Track, TrackList, TracksConfig } from '../../interfaces';

export type TracksListComponentProps = {
  options: TracksConfig;
};

export const TracksListComponent = ({ options }: TracksListComponentProps) => {
  const tracks = useApiGet<Track[]>({
    endpoint: options.API_ROUTE_TRACKS,
  });

  return (
    <div className="mt40">
      {tracks &&
        tracks
          .map(transformTrackToEnhanced)
          .reduce(groupTracksByMonth.bind(null, options), [])
          .sort(orderGroupByKeyAsc)
          .map(([key, { label, tracks: items }]: TrackList) => (
            <TracksGroupComponent
              key={key}
              label={label}
              tracks={items}
              options={options}
            />
          ))}
    </div>
  );
};
