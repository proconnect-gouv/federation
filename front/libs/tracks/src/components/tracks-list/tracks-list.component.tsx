import { useApiGet } from '@fc/common';

import { Track, TrackList, TracksConfig } from '../../interfaces';
import { groupTracksByMonth, orderGroupByKeyAsc, transformTrackToEnhanced } from '../../utils';
import { TracksGroupComponent } from './tracks-group';

export type TracksListComponentProps = {
  options: TracksConfig;
};

export const TracksListComponent = ({ options }: TracksListComponentProps) => {
  const tracks = useApiGet<Track[]>({
    endpoint: options.API_ROUTE_TRACKS,
  });

  return (
    <div className="fr-mt-5w" id="tracks-list">
      {tracks &&
        tracks
          .map(transformTrackToEnhanced)
          .reduce(groupTracksByMonth(options), [])
          .sort(orderGroupByKeyAsc)
          .map(([tracksGroupKey, { label, tracks: items }]: TrackList) => (
            <TracksGroupComponent
              key={tracksGroupKey}
              label={label}
              options={options}
              tracks={items}
            />
          ))}
    </div>
  );
};
