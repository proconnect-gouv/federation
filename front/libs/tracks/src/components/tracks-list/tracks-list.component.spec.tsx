import { render } from '@testing-library/react';
import { DateTime } from 'luxon';
import { mocked } from 'ts-jest/utils';

import { useApiGet } from '@fc/common';

import { orderGroupByKeyAsc, transformTrackToEnhanced } from '../../utils';
import { TracksGroupComponent } from './tracks-group';
import { TracksListComponent } from './tracks-list.component';

jest.mock('@fc/common');
jest.mock('./tracks-group');
jest.mock('./../../utils/tracks.util');

const tracks = [
  {
    accountId: 'mock-accountId-1',
    city: 'mock-city-1',
    country: 'mock-country-1',
    date: '2021-10-05T14:48:00.000Z',
    datetime: DateTime.fromISO('2021-10-05T14:48:00.000Z'),
    event: 'mock-event-1',
    spAcr: 'mock-spacr-1',
    spId: 'mock-spid-1',
    spName: 'mock-spname-1',
    trackId: 'mock-trackid-1',
  },
  {
    accountId: 'mock-accountId-2',
    city: 'mock-city-2',
    country: 'mock-country-2',
    date: '2021-11-05T14:48:00.000Z',
    datetime: DateTime.fromISO('2021-11-05T14:48:00.000Z'),
    event: 'mock-event-2',
    spAcr: 'mock-spacr-2',
    spId: 'mock-spid-2',
    spName: 'mock-spname-2',
    trackId: 'mock-trackid-2',
  },
];

describe('TracksListComponent', () => {
  const useApiGetMock = mocked(useApiGet);
  useApiGetMock.mockReturnValue(tracks);

  const options = {
    API_ROUTE_TRACKS: 'mock_API_ROUTE_TRACKS',
    API_ROUTE_USER_INFOS: 'mock_API_ROUTE_USER_INFOS',
    LUXON_FORMAT_DAY: 'DDD',
    LUXON_FORMAT_HOUR_MINS: 'T',
    LUXON_FORMAT_MONTH_YEAR: 'LLLL yyyy',
    LUXON_FORMAT_TIMEZONE: 'z',
  };

  it('should have called useApiGet hook', () => {
    // given
    render(<TracksListComponent options={options} />);
    // then
    expect(useApiGetMock).toHaveBeenCalled();
  });

  it('should have called transformTrackToEnhanced', () => {
    // given
    render(<TracksListComponent options={options} />);
    // then
    expect(transformTrackToEnhanced).toHaveBeenCalled();
    expect(transformTrackToEnhanced).toHaveBeenNthCalledWith(1, tracks[0], 0, tracks);
    expect(transformTrackToEnhanced).toHaveBeenNthCalledWith(2, tracks[1], 1, tracks);
  });

  it('should have called orderGroupByKeyAsc', () => {
    // given
    render(<TracksListComponent options={options} />);
    // then
    expect(orderGroupByKeyAsc).toHaveBeenCalled();
  });

  it('should have called TracksGroupComponent', () => {
    // given
    render(<TracksListComponent options={options} />);
    // then
    expect(TracksGroupComponent).toHaveBeenCalled();
    expect(TracksGroupComponent).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        label: 'Octobre 2021',
        options,
        tracks: [tracks[0]],
      }),
      {},
    );
    expect(TracksGroupComponent).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        label: 'Novembre 2021',
        options,
        tracks: [tracks[1]],
      }),
      {},
    );
  });
});
