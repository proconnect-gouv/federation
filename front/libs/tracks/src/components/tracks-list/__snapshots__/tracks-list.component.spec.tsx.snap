// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TracksListComponent should match snapshot 1`] = `
<div>
  <div
    class="fr-mt-5w"
    id="tracks-list"
  >
    <div>
      FooBar TracksGroupComponent
    </div>
    <div>
      FooBar TracksGroupComponent
    </div>
    <div
      class="flex-columns flex-center"
      id="tracks-pagination"
    >
      <div>
        PaginationComponent
      </div>
    </div>
  </div>
</div>
`;

exports[`TracksListComponent should match snapshot when tracks data are not defined 1`] = `
<div>
  <div
    class="fr-mt-5w"
    id="tracks-list"
  >
    <div
      class="flex-columns flex-center"
      id="tracks-pagination"
    />
  </div>
</div>
`;
